File: go/code_smells/oop_abusers/switch_statements.go
package oop_abusers

import "fmt"

// PaymentMethod represents different payment methods
type PaymentMethod string

const (
	CreditCard PaymentMethod = "credit_card"
	PayPal     PaymentMethod = "paypal"
	BankTransfer PaymentMethod = "bank_transfer"
	Cash       PaymentMethod = "cash"
)

// PaymentProcessor demonstrates switch statement code smell
type PaymentProcessor struct{}

// ProcessPayment processes a payment using switch statements
func (pp *PaymentProcessor) ProcessPayment(method PaymentMethod, amount float64) error {
	switch method {
	case CreditCard:
		fmt.Printf("Processing credit card payment for $%.2f\n", amount)
		fmt.Println("Validating card...")
		fmt.Println("Contacting payment gateway...")
		fmt.Println("Payment successful")
		return nil

	case PayPal:
		fmt.Printf("Processing PayPal payment for $%.2f\n", amount)
		fmt.Println("Redirecting to PayPal...")
		fmt.Println("Payment successful")
		return nil

	case BankTransfer:
		fmt.Printf("Processing bank transfer for $%.2f\n", amount)
		fmt.Println("Generating transfer reference...")
		fmt.Println("Payment pending confirmation")
		return nil

	case Cash:
		fmt.Printf("Processing cash payment for $%.2f\n", amount)
		fmt.Println("Payment received")
		return nil

	default:
		return fmt.Errorf("unknown payment method: %s", method)
	}
}

// CalculateFee calculates processing fee - another switch on the same type
func (pp *PaymentProcessor) CalculateFee(method PaymentMethod, amount float64) float64 {
	switch method {
	case CreditCard:
		return amount * 0.029 // 2.9% for credit cards
	case PayPal:
		return amount * 0.034 // 3.4% for PayPal
	case BankTransfer:
		return 2.5 // flat fee for bank transfers
	case Cash:
		return 0 // no fee for cash
	default:
		return 0
	}
}

// GetPaymentMethodName gets the display name - yet another switch
func (pp *PaymentProcessor) GetPaymentMethodName(method PaymentMethod) string {
	switch method {
	case CreditCard:
		return "Tarjeta de Crédito"
	case PayPal:
		return "PayPal"
	case BankTransfer:
		return "Transferencia Bancaria"
	case Cash:
		return "Efectivo"
	default:
		return "Desconocido"
	}
}

// Exercise: Refactor this code to use polymorphism instead of switch statements
// 1. Create a PaymentStrategy interface with methods: Process, CalculateFee, GetName
// 2. Implement the interface for each payment method
// 3. Replace the switch statements with polymorphic calls
// 4. Notice how adding a new payment method becomes easier

File: java/src/main/java/com/refactoring/codesmells/oopabusers/SwitchStatements.java
package com.refactoring.codesmells.oopabusers;

/**
 * Example of Switch Statements code smell.
 */
public class SwitchStatements {

    public enum PaymentMethod {
        CREDIT_CARD,
        PAYPAL,
        BANK_TRANSFER,
        CASH
    }

    public static class PaymentProcessor {

        public void processPayment(PaymentMethod method, double amount) {
            switch (method) {
                case CREDIT_CARD:
                    System.out.printf("Processing credit card payment for $%.2f%n", amount);
                    System.out.println("Validating card...");
                    System.out.println("Contacting payment gateway...");
                    System.out.println("Payment successful");
                    break;

                case PAYPAL:
                    System.out.printf("Processing PayPal payment for $%.2f%n", amount);
                    System.out.println("Redirecting to PayPal...");
                    System.out.println("Payment successful");
                    break;

                case BANK_TRANSFER:
                    System.out.printf("Processing bank transfer for $%.2f%n", amount);
                    System.out.println("Generating transfer reference...");
                    System.out.println("Payment pending confirmation");
                    break;

                case CASH:
                    System.out.printf("Processing cash payment for $%.2f%n", amount);
                    System.out.println("Payment received");
                    break;

                default:
                    throw new IllegalArgumentException("Unknown payment method: " + method);
            }
        }

        // Another switch on the same type
        public double calculateFee(PaymentMethod method, double amount) {
            switch (method) {
                case CREDIT_CARD:
                    return amount * 0.029; // 2.9% for credit cards
                case PAYPAL:
                    return amount * 0.034; // 3.4% for PayPal
                case BANK_TRANSFER:
                    return 2.5; // flat fee for bank transfers
                case CASH:
                    return 0; // no fee for cash
                default:
                    return 0;
            }
        }

        // Yet another switch
        public String getPaymentMethodName(PaymentMethod method) {
            switch (method) {
                case CREDIT_CARD:
                    return "Tarjeta de Crédito";
                case PAYPAL:
                    return "PayPal";
                case BANK_TRANSFER:
                    return "Transferencia Bancaria";
                case CASH:
                    return "Efectivo";
                default:
                    return "Desconocido";
            }
        }
    }

    /*
     * Exercise: Refactor this code to use polymorphism instead of switch statements
     * 1. Create a PaymentStrategy interface with methods: process, calculateFee, getName
     * 2. Implement the interface for each payment method
     * 3. Replace the switch statements with polymorphic calls
     * 4. Notice how adding a new payment method becomes easier
     */
}
